docs/source/developer_guide/testing_framework.rst

```diff
@@ -0,0 +1,269 @@
+ For instance, ``forecasting/tests/test_all_forecasters.py``, or ``distances/tests/test_all_dist_kernels.py``.
+* Learning task specific tests should not duplicate package level, generic estimator tests in ``test_all_estimators.py``
+
+Test code architecture
+----------------------
+
+.. _pytestuse: https://docs.pytest.org/en/6.2.x/example/index.html
+
+``sktime`` test files should use best ``pytest`` practice such as fixtures or test parameterization where possible,
+instead of custom logic, see `pytest documentation <pytestuse>`_.
+
+Estimator tests use ``sktime``'s framework plug-in to ``pytest_generate_tests``,
+which parameterizes estimator fixtures and data input scenarios.
+
+An illustrative example
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Starting with an example:
+
+.. code-block::
+
+   def test_fit_returns_self(estimator_instance, scenario):
+      """Check that fit returns self."""
+      fit_return = scenario.run(estimator_instance, method_sequence=["fit"])
+      assert (
+         fit_return is estimator_instance
+      ), f"Estimator: {estimator_instance} does not return self when calling fit"
+
+Here, ``estimator_instance`` and ``scenario`` fixture are looped over, assuming
+values from a list of fixtures:
+* ``estimator_instance`` over estimator instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``
+* ``scenario`` objects, which encodes data inputs and method call sequences to ``estimator_instance`` (explained in further detail below).
+
+The ``scenario.run`` command is equivalent to calling ``estimator_instance.fit(**scenario_kwargs)``,
+where the ``scenario_kwargs`` are generated by the ``scenario``.
+
+It should be noted that the test is not decorated with fixture parametrization,
+the fixtures are instead generated by ``pytest_generate_tests``.
+
+The reason for this is that the applicable scenarios (fixture values of ``scenario``) depend on the ``estimator_instance`` fixture,
+since inputs to ``fit`` of a classifier will differ to an input to ``fit`` of a forecaster.
+
+Parameterized fixtures
+~~~~~~~~~~~~~~~~~~~~~~
+
+.. _pytestfixtparam: https://docs.pytest.org/en/6.2.x/parametrize.html
+
+``sktime`` uses ``pytest`` fixture parameterization to execute tests in a loop over fixtures,
+for instance running all interface compatibility tests for all estimators.
+See the `pytest documentation on fixture parameterization <pytestfixtparam>`_ for an explanation of fixture parameterization.
+
+Currently, the ``sktime`` testing framework parameterizes the following fixtures in module level tests:
+
+* ``estimator``: all estimator classes, inheriting from the base class of the given module.
+In ``test_all_estimators``, this loops over all estimators.
+* ``estimator_instance``: all estimator test instances, obtained from all ``sktime`` estimators via ``create_test_instances_and_names``
```

Comment:
should I read "estimator test instances" as "estimator instances" or "test instances"?

Reply:
well, the test instances are estimator instances, and they are returned by `create_test_instances_and_names`. Does that help, is it unclear?

Reply:
```suggestion
* ``estimator_instance``: all estimator instances from  ``create_test_instances_and_names``
```

Reply:
that is less clear (this is a method, not a function!), so I think this is worse than the original.

Reply:
I don't think I understand what you mean in that sentence and in your comment, up to you to leave it as is